prac_07/project_management.py

```diff
@@ -0,0 +1,138 @@
+from project import Project
+
+
+def main():
+    projects = []
+    filename = "projects.txt"  # Change this filename if needed
+
+    # Load projects from the data file
+    projects = load_projects(filename)
+
+    choice = ""
+    while choice != "Q":
+        print("- (L)oad projects")
+        print("- (S)ave projects")
+        print("- (D)isplay projects")
+        print("- (F)ilter projects by date")
+        print("- (A)dd new project")
+        print("- (U)pdate project")
+        print("- (Q)uit")
+        choice = input(">>> ").strip().lower()
+
+        if choice == 'l':
+            filename = input("Enter filename to load projects from: ")
+            projects = load_projects(filename)
+
+        elif choice == 's':
+            filename = input("Enter filename to save projects to: ")
+            save_projects(filename, projects)
+
+        elif choice == 'd':
+            display_projects(projects)
+
+        elif choice == 'f':
+            start_date = datetime.datetime.strptime(input("Show projects that start after date (dd/mm/yy): "), "%d/%m/%Y").date()
+            filter_projects_by_date(projects, start_date)
+
+        elif choice == 'a':
+            add_new_project(projects)
+
+        elif choice == 'u':
+            update_project(projects)
+
+        elif choice == 'q':
+            break
+
+        else:
+            print("Invalid choice. Please try again.")
+
+    print("Thank you for using custom-built project management software.")
+
+
+def load_projects(filename):
+    projects = []
+    try:
+        with open(filename, "r") as file:
+            lines = file.readlines()
+            if lines and lines[0].strip() == "Name\tStart Date\tPriority\tCost Estimate\tCompletion Percentage":  # Check if the first line is the header
+                lines = lines[1:]  # Skip the header line
+            for line in lines:
+                project_data = line.strip().split("\t")  # Use tabs ("\t") as the separator
+                if len(project_data) == 5:  # Ensure all fields are present
+                    name = project_data[0]
+                    start_date = datetime.datetime.strptime(project_data[1], "%d/%m/%Y").date()
+                    priority = int(project_data[2])
+                    estimate = float(project_data[3])
+                    completion = int(project_data[4])
+                    projects.append(Project(name, start_date, priority, estimate, completion))
+                else:
+                    print(f"Error reading project data: {line.strip()}")
+    except FileNotFoundError:
+        print("File not found. Starting with an empty project list.")
+    return projects
+
+
+def save_projects(filename, projects):
+    with open(filename, "w") as file:
+        for project in projects:
+            file.write(f"{project.name},{project.start_date.strftime('%d/%m/%Y')},{project.priority},"
+                       f"{project.estimate},{project.completion}\n")
+
+
+def display_projects(projects):
```

Comment:
This is really readable. When I wrote my version, I felt constrained not to create and store new lists if possible, and just used the __lt__ method defined in the Project class to sort. Instead of storing new lists, I looped through the projects list twice, once to print the incomplete projects, and then again for the complete projects, using an `if not project.is_complete()` condition to get the former, and an `if project.is_complete()` condition to get the latter.

I also tied myself in knots, trying to get the date output in the display to match the supplied sample code formatting (dd/mm/yyyy)!